File: kotlin/src/test/kotlin/aethernet/security/Bip39FixtureTest.kt
// SPDX-License-Identifier: MIT
package aethernet.security

import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.File
import java.security.SecureRandom
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Cross-language BIP-39 fixture verifier and end-to-end recovery-phrase
 * exercises.
 *
 * Verifies that the Kotlin implementation reproduces the official Trezor BIP-39
 * test vectors (fixtures/bip39/vectors.json) byte-for-byte — the same three
 * columns (entropy -> mnemonic -> seed) that every AetherNet language SDK and
 * the C# reference must reproduce. Any drift surfaces here as a hex mismatch.
 */
class Bip39FixtureTest {

    // ─── Official Trezor vectors (the parity gate) ──────────────────────────

    @Test
    fun `bip39 official trezor vectors round-trip byte-for-byte`() {
        val (passphrase, vectors) = loadVectors()
        assertEquals("TREZOR", passphrase)
        assertEquals(24, vectors.size)

        for (v in vectors) {
            val entropyHex = v["entropy"]!!
            val mnemonic = v["mnemonic"]!!
            val seedHex = v["seed"]!!

            // entropy -> mnemonic
            assertEquals(mnemonic, Bip39.entropyToMnemonic(unhex(entropyHex)),
                "entropyToMnemonic mismatch for $entropyHex")

            // mnemonic -> entropy (checksum enforced)
            assertEquals(entropyHex, hex(Bip39.mnemonicToEntropy(mnemonic)),
                "mnemonicToEntropy mismatch for '$mnemonic'")

            // mnemonic -> 64-byte seed (PBKDF2-HMAC-SHA512, passphrase TREZOR)
            assertEquals(seedHex, hex(Bip39.mnemonicToSeed(mnemonic, passphrase)),
                "mnemonicToSeed mismatch for '$mnemonic'")
        }
    }

    // ─── (a) Identity backup: known 24-word vector ──────────────────────────

    @Test
    fun `identity recovery phrase matches known vector and restores private key`() {
        val entropyHex = "f585c11aec520db57dd353c69554b21a89b20fb0650966fa0a9d6f74fd989d8f"
        val expectedPhrase =
            "void come effort suffer camp survey warrior heavy shoot primary clutch crush " +
            "open amazing screen patrol group space point ten exist slush involve unfold"

        val seed = unhex(entropyHex)
        assertEquals(expectedPhrase, Bip39.toRecoveryPhrase(seed))

        val (restoredPrivate, _) = Bip39.fromRecoveryPhrase(expectedPhrase)
        assertEquals(entropyHex, hex(restoredPrivate))
    }

    // ─── (b) Random identity: backup -> restore -> sign/verify ──────────────

    @Test
    fun `random identity survives backup and restore with working signature`() {
        val rng = SecureRandom()
        val seed = ByteArray(32).also { rng.nextBytes(it) }
        // Independent oracle for the expected public key (BouncyCastle directly).
        val expectedPublic = Ed25519PrivateKeyParameters(seed, 0).generatePublicKey().encoded

        val phrase = Bip39.toRecoveryPhrase(seed)
        assertEquals(24, phrase.split(' ').size)
        assertTrue(Bip39.isValid(phrase))

        val (restoredPrivate, restoredPublic) = Bip39.fromRecoveryPhrase(phrase)
        assertContentEquals(seed, restoredPrivate, "restored private key must match")
        assertContentEquals(expectedPublic, restoredPublic, "restored public key must match")

        // The restored key is a live Ed25519 identity: it signs and verifies.
        val message = "the mesh is alive".toByteArray()
        val signature = Ed25519Service.sign(restoredPrivate, message)
        assertTrue(Ed25519Service.verify(restoredPublic, message, signature),
            "restored identity must produce a verifiable signature")
    }

    // ─── (c) Reject paths: a mistyped phrase must throw, never silently wrong ─

    @Test
    fun `bad checksum phrase is rejected`() {
        // 24 x "abandon" is well-formed structurally but has an invalid checksum.
        val badChecksum = (0 until 24).joinToString(" ") { "abandon" }
        assertThrows<IllegalArgumentException> { Bip39.mnemonicToEntropy(badChecksum) }
        assertThrows<IllegalArgumentException> { Bip39.fromRecoveryPhrase(badChecksum) }
        assertFalse(Bip39.isValid(badChecksum))
    }

    @Test
    fun `unknown word is rejected`() {
        // Valid 24-word phrase with one word swapped for a non-wordlist token.
        val valid =
            "void come effort suffer camp survey warrior heavy shoot primary clutch crush " +
            "open amazing screen patrol group space point ten exist slush involve unfold"
        val words = valid.split(' ').toMutableList()
        words[5] = "notaword"
        val bad = words.joinToString(" ")
        assertThrows<IllegalArgumentException> { Bip39.mnemonicToEntropy(bad) }
        assertFalse(Bip39.isValid(bad))
    }

    @Test
    fun `wrong word count is rejected`() {
        val threeWords = "abandon ability able"
        assertThrows<IllegalArgumentException> { Bip39.mnemonicToEntropy(threeWords) }
        assertThrows<IllegalArgumentException> { Bip39.fromRecoveryPhrase(threeWords) }
        assertFalse(Bip39.isValid(threeWords))
    }

    @Test
    fun `non-32-byte entropy is rejected for identity backup`() {
        // A valid 12-word phrase decodes to 16 bytes — not a 256-bit identity seed.
        val twelveWord =
            "abandon abandon abandon abandon abandon abandon " +
            "abandon abandon abandon abandon abandon about"
        assertTrue(Bip39.isValid(twelveWord)) // valid BIP-39, just not an identity
        assertThrows<IllegalArgumentException> { Bip39.fromRecoveryPhrase(twelveWord) }

        assertThrows<IllegalArgumentException> { Bip39.toRecoveryPhrase(ByteArray(16)) }
    }

    // ─── Wordlist integrity ─────────────────────────────────────────────────

    @Test
    fun `embedded wordlist has 2048 words matching the fixture`() {
        val fixtureWords = File(repoRoot(), "fixtures/bip39/english.txt")
            .readLines()
            .filter { it.isNotBlank() }
        assertEquals(2048, Bip39.words.size)
        assertEquals(fixtureWords, Bip39.words, "embedded wordlist must equal fixtures/bip39/english.txt")
    }

    // ─── Helpers (mirrors SignalFixtureTest) ────────────────────────────────

    private fun loadVectors(): Pair<String, List<Map<String, String>>> {
        val json = File(repoRoot(), "fixtures/bip39/vectors.json").readText()
        val passphrase = Regex("\"passphrase\"\\s*:\\s*\"([^\"]*)\"")
            .find(json)!!.groupValues[1]
        val vectors = parseArray(json, "vectors")
        return passphrase to vectors
    }

    private fun repoRoot(): File {
        var dir: File? = File(".").canonicalFile
        repeat(10) {
            val candidate = File(dir, "AetherNetProtocol.slnx")
            if (candidate.exists()) return dir!!
            dir = dir?.parentFile ?: return@repeat
        }
        throw IllegalStateException("AetherNetProtocol.slnx not found from ${File(".").canonicalFile}")
    }

    /** Minimal JSON parser for our fixed-shape fixture file (named array of string->string objects). */
    private fun parseArray(json: String, arrayName: String): List<Map<String, String>> {
        val nameIdx = json.indexOf("\"$arrayName\"")
        require(nameIdx >= 0) { "No '$arrayName' array in vectors.json" }
        val openBracket = json.indexOf('[', nameIdx)
        var depth = 0
        var endIdx = -1
        for (i in openBracket until json.length) {
            when (json[i]) {
                '[' -> depth++
                ']' -> { depth--; if (depth == 0) { endIdx = i; break } }
            }
        }
        require(endIdx > 0) { "Unbalanced '$arrayName' array" }
        val arrayBody = json.substring(openBracket + 1, endIdx)
        return splitObjects(arrayBody).map { parseFlatJson(it) }
    }

    private fun splitObjects(text: String): List<String> {
        val result = mutableListOf<String>()
        var depth = 0
        var start = -1
        for (i in text.indices) {
            when (text[i]) {
                '{' -> { if (depth == 0) start = i; depth++ }
                '}' -> { depth--; if (depth == 0 && start >= 0) { result += text.substring(start, i + 1); start = -1 } }
            }
        }
        return result
    }

    private fun parseFlatJson(json: String): Map<String, String> {
        val result = mutableMapOf<String, String>()
        val regex = Regex("\"([A-Za-z_0-9\\$]+)\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"")
        for (match in regex.findAll(json)) {
            result[match.groupValues[1]] = match.groupValues[2]
        }
        return result
    }

    private fun unhex(s: String): ByteArray {
        require(s.length % 2 == 0) { "hex string must have even length" }
        return ByteArray(s.length / 2) { i ->
            ((Character.digit(s[i * 2], 16) shl 4) + Character.digit(s[i * 2 + 1], 16)).toByte()
        }
    }

    private fun hex(b: ByteArray): String =
        b.joinToString("") { "%02x".format(it.toInt() and 0xFF) }
}

File: go/security/bip39_fixture_test.go
// SPDX-License-Identifier: MIT

package security

import (
	"bytes"
	"crypto/ed25519"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// Cross-language BIP-39 fixture verifier. Drives the recovery-phrase codec
// against the official Trezor test vectors in fixtures/bip39/vectors.json and
// asserts entropy->mnemonic->seed byte-for-byte, then exercises the AetherNet
// identity backup/restore round-trip and the reject paths.
//
// Any drift between this Go implementation and the C# reference (or any other
// language) shows up here as a mismatch.

type bip39Vectors struct {
	Passphrase string `json:"passphrase"`
	Vectors    []struct {
		Entropy  string `json:"entropy"`
		Mnemonic string `json:"mnemonic"`
		Seed     string `json:"seed"`
	} `json:"vectors"`
}

func TestBip39Fixture_TrezorVectors(t *testing.T) {
	raw, err := os.ReadFile(findBip39Fixture(t))
	if err != nil {
		t.Fatal(err)
	}
	var doc bip39Vectors
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatal(err)
	}
	if len(doc.Vectors) != 24 {
		t.Fatalf("expected 24 vectors in vectors.json, got %d", len(doc.Vectors))
	}
	if doc.Passphrase != "TREZOR" {
		t.Fatalf("expected passphrase TREZOR, got %q", doc.Passphrase)
	}

	for i, v := range doc.Vectors {
		entropy := mustHexBip39(t, v.Entropy)

		// entropy -> mnemonic
		gotMnemonic, err := EntropyToMnemonic(entropy)
		if err != nil {
			t.Fatalf("vector %d: EntropyToMnemonic error: %v", i, err)
		}
		if gotMnemonic != v.Mnemonic {
			t.Errorf("vector %d: mnemonic mismatch:\n  expected: %s\n  actual:   %s", i, v.Mnemonic, gotMnemonic)
		}

		// mnemonic -> entropy
		gotEntropy, err := MnemonicToEntropy(v.Mnemonic)
		if err != nil {
			t.Fatalf("vector %d: MnemonicToEntropy error: %v", i, err)
		}
		if got := hex.EncodeToString(gotEntropy); got != v.Entropy {
			t.Errorf("vector %d: entropy mismatch:\n  expected: %s\n  actual:   %s", i, v.Entropy, got)
		}

		// mnemonic -> seed (PBKDF2-HMAC-SHA512, passphrase TREZOR)
		gotSeed := MnemonicToSeed(v.Mnemonic, doc.Passphrase)
		if got := hex.EncodeToString(gotSeed); got != v.Seed {
			t.Errorf("vector %d: seed mismatch:\n  expected: %s\n  actual:   %s", i, v.Seed, got)
		}
	}
}

// TestBip39Identity_KnownVector pins the AetherNet identity mapping to a fixed
// 32-byte seed (last Trezor vector): the seed must encode to the exact 24-word
// phrase, and restoring that phrase must recover the same private seed.
func TestBip39Identity_KnownVector(t *testing.T) {
	const (
		entropyHex   = "f585c11aec520db57dd353c69554b21a89b20fb0650966fa0a9d6f74fd989d8f"
		wantMnemonic = "void come effort suffer camp survey warrior heavy shoot primary clutch crush open amazing screen patrol group space point ten exist slush involve unfold"
	)
	seed := mustHexBip39(t, entropyHex)

	phrase, err := ToRecoveryPhrase(seed)
	if err != nil {
		t.Fatalf("ToRecoveryPhrase error: %v", err)
	}
	if phrase != wantMnemonic {
		t.Errorf("recovery phrase mismatch:\n  expected: %s\n  actual:   %s", wantMnemonic, phrase)
	}

	_, priv, err := FromRecoveryPhrase(wantMnemonic)
	if err != nil {
		t.Fatalf("FromRecoveryPhrase error: %v", err)
	}
	if !bytes.Equal(priv, seed) {
		t.Errorf("restored private key mismatch:\n  expected: %s\n  actual:   %s", entropyHex, hex.EncodeToString(priv))
	}
}

// TestBip39Identity_RandomRoundTrip generates a fresh identity, backs it up to a
// phrase, restores it, and confirms the restored key pair is identical and still
// signs+verifies — i.e. the phrase truly is the identity.
func TestBip39Identity_RandomRoundTrip(t *testing.T) {
	seed := make([]byte, 32)
	if _, err := rand.Read(seed); err != nil {
		t.Fatalf("rand: %v", err)
	}

	phrase, err := ToRecoveryPhrase(seed)
	if err != nil {
		t.Fatalf("ToRecoveryPhrase error: %v", err)
	}

	pub, priv, err := FromRecoveryPhrase(phrase)
	if err != nil {
		t.Fatalf("FromRecoveryPhrase error: %v", err)
	}
	if !bytes.Equal(priv, seed) {
		t.Fatalf("restored private key != original seed")
	}

	// The restored public key must be the canonical Ed25519 public key for this
	// seed (trailing 32 bytes of the expanded private key), i.e. what any Ed25519
	// implementation — including Ed25519Service — derives.
	wantPub := ed25519.NewKeyFromSeed(seed)[32:]
	if !bytes.Equal(pub, wantPub) {
		t.Fatalf("restored public key mismatch:\n  expected: %s\n  actual:   %s",
			hex.EncodeToString(wantPub), hex.EncodeToString(pub))
	}

	// And the restored key actually signs and verifies through Ed25519Service.
	es := NewEd25519Service()
	msg := []byte("the mesh is alive")
	sig, err := es.Sign(priv, msg)
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	if !es.Verify(pub, msg, sig) {
		t.Fatalf("restored key failed to sign+verify")
	}
}

// TestBip39_RejectPaths confirms a mistyped or malformed phrase is rejected
// rather than silently yielding a wrong secret.
func TestBip39_RejectPaths(t *testing.T) {
	cases := []struct {
		name     string
		mnemonic string
	}{
		{
			name:     "bad checksum (24x abandon)",
			mnemonic: "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon",
		},
		{
			name:     "unknown word",
			mnemonic: "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon notaword",
		},
		{
			name:     "wrong word count (3 words)",
			mnemonic: "abandon abandon abandon",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := MnemonicToEntropy(tc.mnemonic); err == nil {
				t.Errorf("MnemonicToEntropy(%q): expected error, got nil", tc.name)
			}
			if _, _, err := FromRecoveryPhrase(tc.mnemonic); err == nil {
				t.Errorf("FromRecoveryPhrase(%q): expected error, got nil", tc.name)
			}
			if IsValidMnemonic(tc.mnemonic) {
				t.Errorf("IsValidMnemonic(%q): expected false", tc.name)
			}
		})
	}
}

// ─── Helpers ─────────────────────────────────────────────────────────────

// findBip39Fixture walks up from the working directory to the repo root (marked
// by AetherNetProtocol.slnx) and returns fixtures/bip39/vectors.json — the same
// walk-up idiom the other Go fixture tests use.
func findBip39Fixture(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	for {
		if _, statErr := os.Stat(filepath.Join(dir, "AetherNetProtocol.slnx")); statErr == nil {
			return filepath.Join(dir, "fixtures", "bip39", "vectors.json")
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatalf("repo root (AetherNetProtocol.slnx) not found walking up from working dir")
		}
		dir = parent
	}
}

func mustHexBip39(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("hex.DecodeString(%q): %v", s, err)
	}
	return b
}
